tinygrad/tensor.py

```diff
@@ -365,6 +365,19 @@ def _pool(self, k_:Tuple[int, ...], stride:Union[Tuple[int, ...], int]=1, dilati
   def avg_pool2d(self, kernel_size=(2,2), stride=None): return self._pool(make_pair(kernel_size), stride if stride is not None else kernel_size).mean(axis=tuple(range(0-len(make_pair(kernel_size)), 0)))
   def max_pool2d(self, kernel_size=(2,2), stride=None): return self._pool(make_pair(kernel_size), stride if stride is not None else kernel_size).max(axis=tuple(range(0-len(make_pair(kernel_size)), 0)))
 
+  def conv_transpose2d(self, weight:Tensor, bias:Optional[Tensor]=None, groups=1, stride=1, dilation=1, padding=0) -> Tensor:
+    HW, trailing = weight.shape[2:], list(range(3, len(weight.shape)+1))
+    x, w = self, weight.reshape(groups, weight.shape[0]//groups, weight.shape[1], *weight.shape[2:]).permute(0,2,1,*trailing).flip(trailing)
+    assert stride == 1, "stride is not supported in transposed conv"
+    # TODO: stride support. i believe this is correct, but you have to SHRINK the output tensor
+    #stride = make_pair(stride, len(HW))
+    #if any(s>1 for s in stride):
+    #  x = x.reshape(bs, cin_, *flatten((k,1) for k in x.shape[2:]))
```

Comment:
This is a terrible common pattern that should be refactored to something einops looking.